### PR TITLE
Basic bignum operations

### PR DESCRIPTION
**Pull request by npwardberkeley**
`add`, `addmul`, `ge`, `iszero`, `mul`, and `shr`

@unzvfu: Sorry if it's hard to compare this with the previous PR, please let me know if I can help at all. But I've implemented all your suggestions, including changing `add` to return the query. The only optimization I haven't done yet is the `mul` optimization to not use scratch space, because I wasn't sure how

I'm also going to add a few more tests before merging

### CODE REVIEW

File: evm/src/cpu/kernel/tests/bignum.rs [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+}
+
+#[test]
+fn test_mul_bignum_all() -> Result<()> {
+    test_mul_bignum(&prepare_two_bignums_random)?;
+    test_mul_bignum(&prepare_two_bignums_max)?;
+    test_mul_bignum(&prepare_two_bignums_min)?;
+    test_mul_bignum(&prepare_two_bignums_diff)?;
+
+    Ok(())
```
Comment (unzvfu) [inline evm/src/cpu/kernel/tests/bignum.rs:375]: I added a function
```
fn prepare_two_bignums_zero(_bit_size: usize) -> (BigUint, BigUint, U256, Vec<U256>) {
    let a = BigUint::zero();
    let b = BigUint::zero();
    let length: U256 = bignum_len(&a).into();
    let memory = pack_bignums(&[a.clone(), b.clone()], length.try_into().unwrap());

    (a, b, length, memory)
}
```
and called it as `test_{add,mul,addmul}_bignum(&prepare_two_bignums_zero)` in the tests above. This causes them to hang (assuming you've made my suggested change to `bignum_len` above), presumably because they don't handle `len == 0`.

**Review by unzvfu — CHANGES_REQUESTED** [review]
Nice work, this is looking nearly done!

The main issue that needs addressing is handling bignums of zero length. Personally I think such a representation should be handled correctly in the code; however, if you decide you don't want/need to, then this decision and the associated caveats need to be much more clearly documented.

It would also be good to see the test suite expanded somewhat. I'd suggest precomputing tables of inputs and expected outputs for each operation, some of which should be chosen deliberately to exercise particular branches in the code, and some a bit more randomly. When I build such tables I like to use [`mpz_random2`](https://gmplib.org/manual/Integer-Random-Numbers#index-mpz_005frandom2) from the GMP library which is designed to generate inputs that are more likely to trigger implementation errors in such code.

File: evm/src/cpu/kernel/asm/bignum/iszero.asm [hunk 2]
```diff
@@ -0,0 +1,36 @@
+// Arithmetic on little-endian integers represented with 128-bit limbs.
+// All integers must be under a given length bound, and are padded with leading zeroes.
+
+global iszero_bignum:
```
[the remark this answers is not in the document]
Comment (npwardberkeley) [reply evm/src/cpu/kernel/asm/bignum/iszero.asm:4]: I'll just add a check!

File: evm/src/cpu/kernel/tests/bignum.rs [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+fn prepare_bignum_random(bit_size: usize) -> (BigUint, U256, Vec<U256>) {
+    let a = gen_bignum(bit_size);
+    let length: U256 = bignum_len(&a).into();
+    let a_limbs = biguint_to_mem_vec(a.clone());
+
+    (a, length, a_limbs)
+}
+
+fn prepare_bignum_max(bit_size: usize) -> (BigUint, U256, Vec<U256>) {
+    let a = BigUint::one() << (bit_size - 1);
```
[the remark this answers is not in the document]
Comment (npwardberkeley) [reply evm/src/cpu/kernel/tests/bignum.rs:52]: Oh good point

File: evm/src/util.rs [hunk 4]
```diff
@@ -144,3 +144,45 @@ pub(crate) fn biguint_to_u256(x: BigUint) -> U256 {
...
+pub(crate) fn mem_vec_to_biguint(x: &[U256]) -> BigUint {
+    le_limbs_to_biguint(&x.iter().map(|&n| n.try_into().unwrap()).collect_vec())
+}
+
+#[cfg(test)]
+pub(crate) fn biguint_to_le_limbs(x: BigUint) -> Vec<u128> {
+    let mut digits = x.to_u32_digits();
+
+    // Pad to a multiple of 8.
+    digits.resize((digits.len() + 7) / 8 * 8, 0);
```
[the remark this answers is not in the document]
Comment (npwardberkeley) [reply evm/src/util.rs:174]: Hm that doesn't work, we want to add the "inverse" of the remainder, not the remainder itself. So we'd want to add `8 - digits.len() % 8`, except that doesn't work in the `digits.len() % 8 == 0` case... do you have an idea to do it more cleanly?

File: evm/src/cpu/kernel/asm/bignum/shr.asm [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    %shr_const(1)
+    // stack: a[i] >> 1, a[i], i, carry, start_loc, retdest
+    SWAP1
+    // stack: a[i], a[i] >> 1, i, carry, start_loc, retdest
+    %mod_const(2)
+    // stack: new_carry = a[i] % 2, a[i] >> 1, i, carry, start_loc, retdest
+    SWAP3
+    // stack: carry, a[i] >> 1, i, new_carry, start_loc, retdest
+    %shl_const(127)
+    // stack: carry << 127, a[i] >> 1, i, new_carry, start_loc, retdest
```
[the remark this answers is not in the document]
Comment (npwardberkeley) [reply evm/src/cpu/kernel/asm/bignum/shr.asm:38]: Ah interesting; I still don't think it's worth switching over at this point but will definitely keep in mind!

File: evm/src/cpu/kernel/asm/bignum/addmul.asm [hunk 6]
```diff
@@ -0,0 +1,99 @@
+// Arithmetic on little-endian integers represented with 128-bit limbs.
+// All integers must be under a given length bound, and are padded with leading zeroes.
+
+// Sets a[0:len] += b[0:len] * val, and returns the carry.
```
[the remark this answers is not in the document]
Comment (npwardberkeley) [reply evm/src/cpu/kernel/asm/bignum/addmul.asm:4]: Hm, can't a carry be a "digit" (limb) too, not just a bit? In grade-school math the carry is a base-10 digit

File: evm/src/cpu/kernel/asm/bignum/mul.asm [hunk 7]
```diff
@@ -0,0 +1,53 @@
+// Arithmetic on little-endian integers represented with 128-bit limbs.
+// All integers must be under a given length bound, and are padded with leading zeroes.
+
+// Stores a * b in output_loc, leaving a and b unchanged.
+// Both a and b have length len; a * b will have length 2 * len.
+// output_loc must be initialized as 2 * len zeroes.
```
[the remark this answers is not in the document]
Comment (npwardberkeley) [reply evm/src/cpu/kernel/asm/bignum/mul.asm:6]: Hmm interesting — but that will make the code more bloated, both in terms of complexity and in terms of kernel size. That initialization will still be complicated because there still needs to be the add-and-carry dynamic from `addmul`. So I'm not sure it's worth it?

File: evm/src/cpu/kernel/asm/bignum/addmul.asm (hunk 6 above)
[the remark this answers is not in the document]
Comment (unzvfu) [reply evm/src/cpu/kernel/asm/bignum/addmul.asm:4]: I overwhelmingly see "carry" to mean "carry bit", especially in low-level and assembly programming, but it's true that is a narrow perspective. The important thing here is to document what the range of the return value is, so feel free to call it simply "carry" but it'd be good to have a few extra words to explain that it is the size of a bignum limb, to avoid any ambiguity.

File: evm/src/cpu/kernel/asm/bignum/mul.asm (hunk 7 above)
[the remark this answers is not in the document]
Comment (unzvfu) [reply evm/src/cpu/kernel/asm/bignum/mul.asm:6]: Yeah, it's a trade-off; the change will indeed increase the kernel size, but it will reduce the number of instructions executed per kernel invocation. But at this stage I agree, probably not worth it, especially if it's not a burden for the calling code to guarantee that the memory is initialised to zero before every call.

File: evm/src/util.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (unzvfu) [reply evm/src/util.rs:174]: Oh right, apologies for the confusion! What you have is fine. :)

File: evm/src/cpu/kernel/asm/bignum/addmul.asm (hunk 6 above)
[the remark this answers is not in the document]
Comment (npwardberkeley) [reply evm/src/cpu/kernel/asm/bignum/addmul.asm]: I changed it to "carry limb" everywhere, I think that's the best of both worlds? So it's clear that it's functioning as the carry but there's no confusion that it's just one bit

File: evm/src/cpu/kernel/asm/bignum/mul.asm (hunk 7 above)
[the remark this answers is not in the document]
Comment (npwardberkeley) [reply evm/src/cpu/kernel/asm/bignum/mul.asm:6]: I'll leave a TODO for now!

File: evm/src/cpu/kernel/evm_asm.pest [hunk 8]
```diff
@@ -26,7 +26,7 @@ stack = { ^"%stack" ~ stack_placeholders ~ "->" ~ stack_replacements }
 stack_placeholders = { "(" ~ (stack_placeholder ~ ("," ~ stack_placeholder)*)? ~ ")" }
 stack_placeholder = { stack_block | identifier }
 stack_block = { identifier ~ ":" ~ literal_decimal }
-stack_replacements = { "(" ~ stack_replacement ~ ("," ~ stack_replacement)* ~ ")" }
+stack_replacements = { "(" ~ (stack_replacement ~ ("," ~ stack_replacement)*)? ~ ")" }
```
Comment (unzvfu) [inline evm/src/cpu/kernel/evm_asm.pest:29]: This doesn't really belong in this PR.

File: plonky2/src/fri/recursive_verifier.rs
Comment (unzvfu) [inline plonky2/src/fri/recursive_verifier.rs:1]: These changes (and the similar format string changes in other files unrelated to the bignum code) should really be in their own PR.

File: evm/src/cpu/kernel/tests/bignum/mod.rs [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+
+    let inputs = test_data_biguint(TEST_DATA_BIGNUM_INPUTS);
+    let cmp_outputs = test_data_u256(TEST_DATA_CMP_OUTPUTS);
+    let mut cmp_outputs_iter = cmp_outputs.iter();
+    for a in inputs.clone() {
+        for b in inputs.clone() {
+            let output = cmp_outputs_iter.next().unwrap();
+            test_cmp_bignum(a.clone(), b.clone(), *output)?;
+        }
+    }
```
Comment (unzvfu) [inline evm/src/cpu/kernel/tests/bignum/mod.rs:303]: Here and in the other tests with similar code, I think you can just iterate over slices rather than cloning the whole input.
```suggestion
    for a in &inputs {
        for b in &inputs {
            let output = cmp_outputs_iter.next().unwrap();
            test_cmp_bignum(a.clone(), b.clone(), *output)?;
        }
    }
```

File: evm/src/cpu/kernel/tests/bignum/mod.rs [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+const TEST_DATA_BIGNUM_INPUTS: &str = "bignum_inputs";
+const TEST_DATA_U128_INPUTS: &str = "u128_inputs";
+const TEST_DATA_SHR_OUTPUTS: &str = "shr_outputs";
+const TEST_DATA_ISZERO_OUTPUTS: &str = "iszero_outputs";
+const TEST_DATA_CMP_OUTPUTS: &str = "cmp_outputs";
+const TEST_DATA_ADD_OUTPUTS: &str = "add_outputs";
+const TEST_DATA_ADDMUL_OUTPUTS: &str = "addmul_outputs";
+const TEST_DATA_MUL_OUTPUTS: &str = "mul_outputs";
+
+const BIT_SIZES_TO_TEST: [usize; 4] = [128, 256, 512, 1000];
```
Comment (unzvfu) [inline evm/src/cpu/kernel/tests/bignum/mod.rs:30]: You should probably include borderline bitsizes as well, like 0, 1, 2, 127, 129, 255, 257, 1023, 1024, 1025, and maybe a much bigger one like 31415. You'll need to adjust `max_bignum` and `test_iszero_bignum_all` (and maybe some others) to accommodate bitsize 0.

**Review by unzvfu — APPROVED** [review]
Okay, just a few small things to finish up, and you're ready to go. Nice work!

Regarding the changes that don't belong in this PR, you can leave them in this time, but in future it's best to create separate PRs for these unrelated changes. In general, many small PRs is always preferable to fewer large PRs.
